// @gathertown/gather-game-common/dist/testHelpers/factories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.factories = exports.getDefaultMockPlayerName = void 0;
const fishery_1 = require("fishery");
const playerUtils_1 = require("../src/playerUtils");
const events_1 = require("../src/public/events");
const getDefaultMockPlayerName = (playerId) => `player ${playerId}`;
exports.getDefaultMockPlayerName = getDefaultMockPlayerName;
const mockPlayer = fishery_1.Factory.define(({ sequence, transientParams }) => {
    const playerId = transientParams.id || sequence;
    return Object.assign(Object.assign({}, (0, playerUtils_1.generateDefaultPlayer)()), { name: (0, exports.getDefaultMockPlayerName)(playerId), x: 1, y: 1 });
});
const mockWireObject = fishery_1.Factory.define(({ params, transientParams: { interactable = true }, sequence }) => {
    var _a;
    return ({
        id: (_a = params.id) !== null && _a !== void 0 ? _a : `wire-object-${sequence.toString()}`,
        type: interactable ? events_1.InteractionEnum_ENUM.EMBEDDED_WEBSITE : events_1.InteractionEnum_ENUM.NONE,
        width: 10,
        height: 10,
        normal: "testObject.png",
        x: 0,
        y: 0,
        offsetX: 0,
        offsetY: 0,
        _tags: [],
    });
});
exports.factories = {
    player: mockPlayer,
    wireObject: mockWireObject,
};
