// @gathertown/gather-game-common/dist/src/GameWsCloseCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameWsCloseCode = void 0;
var GameWsCloseCode;
(function (GameWsCloseCode) {
    GameWsCloseCode[GameWsCloseCode["NORMAL_CLOSURE"] = 1000] = "NORMAL_CLOSURE";
    GameWsCloseCode[GameWsCloseCode["GOING_AWAY"] = 1001] = "GOING_AWAY";
    GameWsCloseCode[GameWsCloseCode["PROTOCOL_ERROR"] = 1002] = "PROTOCOL_ERROR";
    GameWsCloseCode[GameWsCloseCode["NO_STATUS_RECEIVED"] = 1005] = "NO_STATUS_RECEIVED";
    GameWsCloseCode[GameWsCloseCode["NO_CLOSE_FRAME"] = 1006] = "NO_CLOSE_FRAME";
    GameWsCloseCode[GameWsCloseCode["INTERNAL_ERROR"] = 1011] = "INTERNAL_ERROR";
    GameWsCloseCode[GameWsCloseCode["SPACE_AT_CAPACITY"] = 1013] = "SPACE_AT_CAPACITY";
    GameWsCloseCode[GameWsCloseCode["CONN_TIMED_OUT"] = 4008] = "CONN_TIMED_OUT";
    GameWsCloseCode[GameWsCloseCode["SPACE_REASSIGNED"] = 4009] = "SPACE_REASSIGNED";
    GameWsCloseCode[GameWsCloseCode["CLIENT_KICKED"] = 4010] = "CLIENT_KICKED";
    GameWsCloseCode[GameWsCloseCode["SPACE_CLOSED"] = 4011] = "SPACE_CLOSED";
    GameWsCloseCode[GameWsCloseCode["UNAUTHORIZED"] = 4003] = "UNAUTHORIZED";
    GameWsCloseCode[GameWsCloseCode["CHILL_OUT"] = 4420] = "CHILL_OUT";
    GameWsCloseCode[GameWsCloseCode["SUS"] = 4666] = "SUS";
    GameWsCloseCode[GameWsCloseCode["HIGHLANDER"] = 4667] = "HIGHLANDER";
})(GameWsCloseCode = exports.GameWsCloseCode || (exports.GameWsCloseCode = {}));
