// @gathertown/gather-game-common/dist/src/sharedConstsForExperiments/RemoteWorkTrivia.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_QUESTION_VALUE = exports.QUESTION_PACKS_INFO = exports.QUESTION_PACKS = exports.TRIVIA_COLORS = exports.TriviaGamePhase = exports.TRIVIA_MAP_ID = void 0;
exports.TRIVIA_MAP_ID = "rw-trivia";
var TriviaGamePhase;
(function (TriviaGamePhase) {
    TriviaGamePhase["UNINITIALIZED"] = "UNINITIALIZED";
    TriviaGamePhase["JOIN"] = "JOIN";
    TriviaGamePhase["SHOW_QUESTION"] = "SHOW_QUESTION";
    TriviaGamePhase["SHOW_ANSWER"] = "SHOW_ANSWER";
    TriviaGamePhase["ENDED"] = "ENDED";
})(TriviaGamePhase = exports.TriviaGamePhase || (exports.TriviaGamePhase = {}));
exports.TRIVIA_COLORS = new Set([
    "#FF9DDE",
    "#FBD54E",
    "#64E9D1",
    "#AEE964",
    "#A966FF",
    "#FC7079",
    "#FF9960",
    "#EA36AC",
    "#B67800",
    "#2D62EC",
    "#37C214",
    "#7230FF",
    "#D42E2E",
    "#E85110",
]);
var QUESTION_PACKS;
(function (QUESTION_PACKS) {
    QUESTION_PACKS["GENERAL_174"] = "general_174";
    QUESTION_PACKS["KOREAN_52"] = "korean_52";
    QUESTION_PACKS["MEDIA_74"] = "media_74";
    QUESTION_PACKS["SPORTS_50"] = "sports_50";
    QUESTION_PACKS["TECH_63"] = "tech_63";
})(QUESTION_PACKS = exports.QUESTION_PACKS || (exports.QUESTION_PACKS = {}));
exports.QUESTION_PACKS_INFO = {
    [QUESTION_PACKS.GENERAL_174]: {
        id: "general_174",
        fileName: "General_174.csv",
        numQuestions: 170,
        categories: [
            "Movies",
            "Music",
            "News Events",
            "Pop Culture",
            "Sports",
            "TV",
            "Video Games",
            "Business",
            "Computers & Technology",
            "Literature",
        ],
        maxDifficulty: 3,
        displayName: "General Trivia",
    },
    [QUESTION_PACKS.KOREAN_52]: {
        id: "korean_52",
        fileName: "Korean_52.csv",
        numQuestions: 50,
        categories: ["Korean Trivia"],
        maxDifficulty: 3,
        displayName: "Korean Trivia",
    },
    [QUESTION_PACKS.MEDIA_74]: {
        id: "media_74",
        fileName: "Media_74.csv",
        numQuestions: 70,
        categories: ["Media Trivia"],
        maxDifficulty: 3,
        displayName: "Media Trivia",
    },
    [QUESTION_PACKS.SPORTS_50]: {
        id: "sports_50",
        fileName: "Sports_50.csv",
        numQuestions: 50,
        categories: ["Sports Trivia"],
        maxDifficulty: 3,
        displayName: "Sports Trivia",
    },
    [QUESTION_PACKS.TECH_63]: {
        id: "tech_63",
        fileName: "Tech_63.csv",
        numQuestions: 60,
        categories: ["Tech Trivia"],
        maxDifficulty: 3,
        displayName: "Tech Trivia",
    },
};
exports.DEFAULT_QUESTION_VALUE = 100;
