// @gathertown/gather-game-common/dist/src/sharedConstsForExperiments/RemoteWorkGrandPrix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUpdatedGrandPrixLeaderboardsData = exports.GRAND_PRIX_MAP_SFX = exports.SoundEffect = exports.BACKGROUND_MUSIC_VOLUME = exports.SOUND_EFFECT_VOLUME = exports.CROWN_IMG = exports.DOOR_AREA_COORDS = exports.AUDIENCE_COORDS = exports.randomIntFromInterval = exports.PODIUM_COORDS = exports.FINISHED_PLACE_COLOR = exports.GRAND_PRIX_CHECKPOINT = exports.GRAND_PRIX_NUM_LANES = exports.GRAND_PRIX_START_POS = exports.GrandPrixClientEvents = exports.GrandPrixServerEvents = exports.GrandPrixGamePhase = exports.NUM_LAPS = exports.GrandPrixMapsInfo = exports.GRAND_PRIX_TEMPLATE_PORTALS = exports.GRAND_PRIX_PORTAL = exports.GRAND_PRIX_MAPS = exports.GrandPrixMaps = void 0;
const RemoteWorkRoomOfRequirements_1 = require("./RemoteWorkRoomOfRequirements");
var GrandPrixMaps;
(function (GrandPrixMaps) {
    GrandPrixMaps["Original"] = "rose-raceway";
    GrandPrixMaps["Castle"] = "dark-castle-dragway";
    GrandPrixMaps["Seaside"] = "seaside-speedway";
})(GrandPrixMaps = exports.GrandPrixMaps || (exports.GrandPrixMaps = {}));
exports.GRAND_PRIX_MAPS = [
    GrandPrixMaps.Original,
    GrandPrixMaps.Castle,
    GrandPrixMaps.Seaside,
];
exports.GRAND_PRIX_PORTAL = RemoteWorkRoomOfRequirements_1.SocialRoomInfo[RemoteWorkRoomOfRequirements_1.SocialRoom.GoKarts].portal;
exports.GRAND_PRIX_TEMPLATE_PORTALS = RemoteWorkRoomOfRequirements_1.SocialRoomInfo[RemoteWorkRoomOfRequirements_1.SocialRoom.GoKarts].templatePortals;
exports.GrandPrixMapsInfo = {
    [GrandPrixMaps.Original]: {
        name: "Rose Raceway",
        imageURL: "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/iYKZr84o97YLlNsd/CMGcE7i6DidMU95T2MBjuQ",
        portal: exports.GRAND_PRIX_PORTAL,
        templatePortals: exports.GRAND_PRIX_TEMPLATE_PORTALS,
        sourceSpace: "iYKZr84o97YLlNsd\\templateRacetrack",
    },
    [GrandPrixMaps.Castle]: {
        name: "Dark Castle Dragway",
        imageURL: "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/FJCjTwKxexpwUUJs/PAK5C3CWKyPp1Kd3CpMpIy",
        portal: exports.GRAND_PRIX_PORTAL,
        templatePortals: exports.GRAND_PRIX_TEMPLATE_PORTALS,
        sourceSpace: "FJCjTwKxexpwUUJs\\templateDarkCastle",
    },
    [GrandPrixMaps.Seaside]: {
        name: "Seaside Speedway",
        imageURL: "https://cdn.gather.town/storage.googleapis.com/gather-town.appspot.com/uploads/HJWM8Cb2JGtgEmyr/iCTyCnMf7OA8fqHNToo5kx",
        portal: exports.GRAND_PRIX_PORTAL,
        templatePortals: exports.GRAND_PRIX_TEMPLATE_PORTALS,
        sourceSpace: "HJWM8Cb2JGtgEmyr\\templateBeachTrack",
    },
};
exports.NUM_LAPS = 3;
var GrandPrixGamePhase;
(function (GrandPrixGamePhase) {
    GrandPrixGamePhase["WAITING_FOR_PLAYERS"] = "waitingForPlayers";
    GrandPrixGamePhase["PLAYING"] = "playing";
    GrandPrixGamePhase["DONE"] = "done";
})(GrandPrixGamePhase = exports.GrandPrixGamePhase || (exports.GrandPrixGamePhase = {}));
var GrandPrixServerEvents;
(function (GrandPrixServerEvents) {
    GrandPrixServerEvents["FULL_STATE"] = "fullState";
    GrandPrixServerEvents["UPDATED_GAME_PHASE"] = "updatedGamePhase";
    GrandPrixServerEvents["FINISHED_LAP"] = "finishedLap";
    GrandPrixServerEvents["PLAYER_FINISHED_RACE"] = "playerFinishedRace";
    GrandPrixServerEvents["ALL_PLAYERS_FINISHED"] = "allPlayersFinished";
    GrandPrixServerEvents["SELECT_TRACK_FINISHED"] = "selectTrackFinished";
    GrandPrixServerEvents["UPDATE_LEADERBOARD"] = "updateLeaderboard";
})(GrandPrixServerEvents = exports.GrandPrixServerEvents || (exports.GrandPrixServerEvents = {}));
var GrandPrixClientEvents;
(function (GrandPrixClientEvents) {
    GrandPrixClientEvents["JOIN_SPACE"] = "joinSpace";
    GrandPrixClientEvents["START_PRESSED"] = "startPressed";
    GrandPrixClientEvents["END_RACE"] = "endRace";
    GrandPrixClientEvents["RESET_RACE"] = "resetRace";
    GrandPrixClientEvents["SELECT_TRACK"] = "selectTrack";
})(GrandPrixClientEvents = exports.GrandPrixClientEvents || (exports.GrandPrixClientEvents = {}));
exports.GRAND_PRIX_START_POS = {
    [GrandPrixMaps.Original]: [39, 73],
    [GrandPrixMaps.Seaside]: [39, 71],
    [GrandPrixMaps.Castle]: [39, 71],
};
exports.GRAND_PRIX_NUM_LANES = {
    [GrandPrixMaps.Original]: 10,
    [GrandPrixMaps.Seaside]: 12,
    [GrandPrixMaps.Castle]: 12,
};
exports.GRAND_PRIX_CHECKPOINT = {
    [GrandPrixMaps.Original]: [46, 7],
    [GrandPrixMaps.Seaside]: [37, 7],
    [GrandPrixMaps.Castle]: [32, 9],
};
exports.FINISHED_PLACE_COLOR = {
    1: "#569cfb",
    2: "#6ece9a",
    3: "#fdef4a",
};
exports.PODIUM_COORDS = {
    1: [48, 53],
    2: [47, 53],
    3: [49, 53],
};
function randomIntFromInterval(min, max) {
    return Math.floor(Math.random() * (max - min + 1) + min);
}
exports.randomIntFromInterval = randomIntFromInterval;
exports.AUDIENCE_COORDS = {
    xMin: 44,
    xMax: 52,
    yMin: 56,
    yMax: 60,
};
exports.DOOR_AREA_COORDS = {
    xMin: 16,
    xMax: 22,
    yMin: 84,
    yMax: 85,
};
exports.CROWN_IMG = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Favatar_goldcrown.png?alt=media&token=adf0c54f-b730-451e-a450-71d3ccd5f897";
exports.SOUND_EFFECT_VOLUME = 0.1;
exports.BACKGROUND_MUSIC_VOLUME = 0.15;
var SoundEffect;
(function (SoundEffect) {
    SoundEffect["MountVehicle"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FUI_select.wav?alt=media&token=e50967bb-20ef-4dea-b4dd-af087c7ecf84";
    SoundEffect["DismountVehicle"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FUI_cancel.mp3?alt=media&token=efdbe39d-ed33-4321-81c2-2cbf34e267b6";
    SoundEffect["Ramp"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FObject_jump.wav?alt=media&token=e6df4e96-5f51-4e1c-a2d7-bff06587d9ce";
    SoundEffect["SpeedBoost"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FObject_booster.wav?alt=media&token=5ce2a956-575e-47d5-8402-cdd2c9dd2fe6";
    SoundEffect["Obstacle"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FObject_obstacle.mp3?alt=media&token=5022a01d-5ed1-4a86-b143-11275f61c7e8";
    SoundEffect["FinishRace"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FFinish%20the%20race.mp3?alt=media&token=e184478e-8e16-4aa4-87d8-a46aff76f735";
    SoundEffect["Countdown"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FCountdown.wav?alt=media&token=0352b662-7696-4612-a1ce-ebcfb7ee1b07";
    SoundEffect["StopOrResetGame"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FStop_Reset_game.mp3?alt=media&token=e61a330f-8bc1-42ce-85d1-9994017a6533";
    SoundEffect["RoseRacewayWaiting"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FRose%20Raceway_waiting_new.mp3?alt=media&token=674e620d-f869-4725-9f76-42a5ce81c238";
    SoundEffect["RoseRacewayGame"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FRose%20Raceway_game_new.mp3?alt=media&token=7ac231cb-8dd2-46b0-82e0-2d74b5472083";
    SoundEffect["BeachWaiting"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FBeach_waiting_new.mp3?alt=media&token=78e16191-b2bf-4017-ac34-758be26390f0";
    SoundEffect["BeachGame"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FBeach_game_new.mp3?alt=media&token=7b8bc128-f9cb-4479-a605-a38c59ef0f32";
    SoundEffect["CastleWaiting"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FCastle_waiting_new.mp3?alt=media&token=b4531367-64a2-4d59-865b-218a83b6e556";
    SoundEffect["CastleGame"] = "https://cdn.gather.town/v0/b/gather-town.appspot.com/o/remote-work-events%2Fgokarts%2Fsfx%2FCastle_game_new.mp3?alt=media&token=994da0b6-8b40-404c-a667-f4618202359f";
})(SoundEffect = exports.SoundEffect || (exports.SoundEffect = {}));
exports.GRAND_PRIX_MAP_SFX = {
    [GrandPrixMaps.Original]: {
        [GrandPrixGamePhase.WAITING_FOR_PLAYERS]: SoundEffect.RoseRacewayWaiting,
        [GrandPrixGamePhase.PLAYING]: SoundEffect.RoseRacewayGame,
        [GrandPrixGamePhase.DONE]: SoundEffect.RoseRacewayGame,
    },
    [GrandPrixMaps.Seaside]: {
        [GrandPrixGamePhase.WAITING_FOR_PLAYERS]: SoundEffect.BeachWaiting,
        [GrandPrixGamePhase.PLAYING]: SoundEffect.BeachGame,
        [GrandPrixGamePhase.DONE]: SoundEffect.BeachGame,
    },
    [GrandPrixMaps.Castle]: {
        [GrandPrixGamePhase.WAITING_FOR_PLAYERS]: SoundEffect.CastleWaiting,
        [GrandPrixGamePhase.PLAYING]: SoundEffect.CastleGame,
        [GrandPrixGamePhase.DONE]: SoundEffect.CastleGame,
    },
};
const getUpdatedGrandPrixLeaderboardsData = (players, mapId, grandPrixLeaderboards, additionalPlayerInfo) => {
    var _a;
    const leaderboardData = [...((_a = grandPrixLeaderboards === null || grandPrixLeaderboards === void 0 ? void 0 : grandPrixLeaderboards[mapId]) !== null && _a !== void 0 ? _a : [])];
    const date = new Date().toISOString();
    const playerDataList = players
        .filter((player) => !player.isSpectator)
        .map(({ playerId, finishedTime }) => {
        var _a, _b;
        return ({
            playerId,
            playerName: (_a = additionalPlayerInfo === null || additionalPlayerInfo === void 0 ? void 0 : additionalPlayerInfo[playerId]) === null || _a === void 0 ? void 0 : _a.name,
            playerOutfitString: (_b = additionalPlayerInfo === null || additionalPlayerInfo === void 0 ? void 0 : additionalPlayerInfo[playerId]) === null || _b === void 0 ? void 0 : _b.outfitString,
            finishedTime,
            date,
        });
    });
    leaderboardData.push(...playerDataList);
    leaderboardData.sort((a, b) => a.finishedTime - b.finishedTime);
    return Object.assign(Object.assign({}, grandPrixLeaderboards), { [mapId]: leaderboardData.slice(0, 10) });
};
exports.getUpdatedGrandPrixLeaderboardsData = getUpdatedGrandPrixLeaderboardsData;
