// @gathertown/gather-game-common/dist/src/sharedConstsForExperiments/RemoteWorkFishing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllJunkItems = exports.getJunkItem = exports.getAllFishItems = exports.getFishItem = exports.getFishingItemNameToImage = exports.JunkName = exports.FishName = void 0;
var FishName;
(function (FishName) {
    FishName["Salmon"] = "Salmon";
    FishName["Trout"] = "Trout";
    FishName["Bass"] = "Bass";
})(FishName = exports.FishName || (exports.FishName = {}));
var JunkName;
(function (JunkName) {
    JunkName["CyberpunkWaste"] = "Cyberpunk Waste";
    JunkName["Anchor"] = "Anchor";
    JunkName["Kelp"] = "Kelp";
})(JunkName = exports.JunkName || (exports.JunkName = {}));
const getFishingItemNameToImage = (name) => `https://cdn.gather.town/v0/b/gather-town.appspot.com/o/manually-uploaded%2Ffishing%2F${name
    .toUpperCase()
    .split(" ")
    .join("_")}.png?alt=media`;
exports.getFishingItemNameToImage = getFishingItemNameToImage;
const getFishItem = (fishName) => {
    const previewUrl = (0, exports.getFishingItemNameToImage)(fishName);
    const item = {
        name: fishName,
        category: "Fishing",
        previewUrl,
        meta: {},
        abilities: {},
    };
    return item;
};
exports.getFishItem = getFishItem;
const getAllFishItems = () => Object.values(FishName).reduce((prev, fishName) => (Object.assign(Object.assign({}, prev), { [fishName]: (0, exports.getFishItem)(fishName) })), {});
exports.getAllFishItems = getAllFishItems;
const getJunkItem = (junkName) => {
    const previewUrl = (0, exports.getFishingItemNameToImage)(junkName);
    const item = {
        name: junkName,
        category: "Fishing Junk",
        previewUrl,
        meta: {},
        abilities: {},
    };
    return item;
};
exports.getJunkItem = getJunkItem;
const getAllJunkItems = () => Object.values(JunkName).reduce((prev, junkName) => (Object.assign(Object.assign({}, prev), { [junkName]: (0, exports.getJunkItem)(junkName) })), {});
exports.getAllJunkItems = getAllJunkItems;
