// @gathertown/gather-game-common/src/public/gameMap.ts
export {
  Area,
  AreaCategory,
  Areas,
  CoordsMap,
  coordsMapToPositions,
  BaseArea,
  convertAreaCoordsToCoordsMap,
  convertCoordsToBBoxes,
  convertDBDesksCoordsToCoordsMap,
  convertDBDesksToDesks,
  convertDesksToDBDesks,
  convertNookInfoToOccupiedNookTilesMap,
  convertMapObjectToWireObject,
  convertStringToEnabledChat,
  convertStringToPresetTutorialGroupSetId,
  convertTutorialTasksToWireTutorialTasks,
  convertWireAreasToAreas,
  convertWireAreasToDBAreas,
  convertWireObjectToMapObject,
  convertWireObjectToMapObjectDB,
  convertWireTutorialTasksToTutorialTasks,
  DBArea,
  DBAreas,
  DBDesks,
  Desks,
  desksToNooks,
  SpawnTile,
  PortalTile,
  TileEffect,
  EnabledChat,
  findMissingMapObjectFields,
  FloorBlock,
  GameMapBaseDoNotUseDirectly,
  GameMapDB,
  GameMapHttpPublic,
  generateEmptyDeskCoordsMap,
  Interaction,
  interactionOpensModal,
  MAP_MAX_DIMENSION,
  MapNameAndId,
  MapObject,
  MapObjectDB,
  ModalInteraction,
  mergeNooks,
  Nook,
  NookInfo,
  NookTile,
  OccupiedNookTilesMap,
  nookToDeprecatedSpaces,
  deprecatedSpacesToNooks,
  PresetTutorialGroupSetIds,
  Tile,
  TileType,
  isTileType,
  TutorialTasks,
  WallBlock,
  WallFloorBlock,
  wallFloorConvertFromDB,
  wallFloorConvertToDB,
  WallFloorDB,
} from "../GameMap";

// We reexport some of the types from the protobuf `events.ts` because this is a more likely and logical place people
// would search for them. Also it makes it possible to auto-import with Intellisense, since it's not possible from
// public/events.ts.
export { Space, Portal, SpawnPoint } from "../generated_DO_NOT_TOUCH/events";
